File: week6/Decorator/Decorator.kt
open class Coffee {
    open fun cost(): Int = 5
}

open class CoffeeDecorator(protected val coffee: Coffee) : Coffee() {
    override fun cost(): Int = coffee.cost()
}

class MilkDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost(): Int = coffee.cost() + 2
}

class SugarDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost(): Int = coffee.cost() + 1
}

fun main() {
    val coffee = Coffee()
    println("Basic coffee: ${coffee.cost()}")

    val coffeeWithMilk = MilkDecorator(coffee)
    println("Coffee with milk: ${coffeeWithMilk.cost()}")

    val coffeeWithSugar = SugarDecorator(coffee)
    println("Coffee with sugar: ${coffeeWithSugar.cost()}")

    val coffeeWithMilkAndSugar = SugarDecorator(coffeeWithMilk)
    println("Coffee with milk and sugar: ${coffeeWithMilkAndSugar.cost()}")
}

File: week6/Decorator/Decorator.cpp
#include <iostream>
using namespace std;

class Coffee {
public:
    virtual int cost() const {
        return 5;
    }
    virtual ~Coffee() = default;
};

class CoffeeDecorator : public Coffee {
protected:
    Coffee* coffee;
public:
    CoffeeDecorator(Coffee* c) : coffee(c) {}
    int cost() const override {
        return coffee->cost();
    }
};

class MilkDecorator : public CoffeeDecorator {
public:
    MilkDecorator(Coffee* c) : CoffeeDecorator(c) {}
    int cost() const override {
        return coffee->cost() + 2;
    }
};

class SugarDecorator : public CoffeeDecorator {
public:
    SugarDecorator(Coffee* c) : CoffeeDecorator(c) {}
    int cost() const override {
        return coffee->cost() + 1;
    }
};

int main() {
    Coffee* coffee = new Coffee();
    cout << "Basic coffee: " << coffee->cost() << endl;

    Coffee* coffeeWithMilk = new MilkDecorator(coffee);
    cout << "Coffee with milk: " << coffeeWithMilk->cost() << endl;

    Coffee* coffeeWithSugar = new SugarDecorator(coffee);
    cout << "Coffee with sugar: " << coffeeWithSugar->cost() << endl;

    Coffee* coffeeWithMilkAndSugar = new SugarDecorator(coffeeWithMilk);
    cout << "Coffee with milk and sugar: " << coffeeWithMilkAndSugar->cost() << endl;

    // 메모리 해제
    delete coffeeWithMilkAndSugar;
    delete coffeeWithSugar;
    delete coffeeWithMilk;
    delete coffee;

    return 0;
}
